### PR TITLE
Poc wf serie

### PR DESCRIPTION
**Pull request by jrmbrgs**
#1

### CODE REVIEW

File: Library/Services/WorkflowManager.php [hunk 1]
```diff
@@ -26,47 +27,262 @@ public function __construct(string $workflowConfigFilePath)
...
     }
 
-    public function finalizeStep(string $workflowProcessId, string $stepCode, array $resultHash) {
+    /**
+     * Check for a step if all related jobs have been succeed
+     *
+     * @param array $step
+     * @return bool
+     */
+    private function didAllStepJobsSucceeded(array $step) : bool
```
Comment (jrmbrgs) [inline Library/Services/WorkflowManager.php]: as seen orally, naming : `$stepHash`
add in comment that this hash comes from the context

File: Library/Services/WorkflowManager.php [hunk 2]
```diff
@@ -26,47 +27,262 @@ public function __construct(string $workflowConfigFilePath)
...
+     *
+     * @param string $workflowProcessId
+     * @return bool
+     */
+    public function isNextStepRunnable(string $workflowProcessId) : bool
+    {
         $this->getDI()->get('logger')->debug(json_encode(func_get_args()));
-        //$currentStepNo = array_search($stepName, array_column($this->config['steps'], 'name'));
+        $runnable = true;
         $currentStepPos = $this->tmpStorage[$workflowProcessId]['currentStepPos'];
```
Comment (jrmbrgs) [inline Library/Services/WorkflowManager.php]: to factorizein a `getWorkflowCurrentPosition(string $workflowProcessId)) : int`

File: Library/Services/WorkflowManager.php [hunk 3]
```diff
@@ -26,47 +27,262 @@ public function __construct(string $workflowConfigFilePath)
...
+        return $succeed;
+    }
+
+    /**
+     * Check current step status and if we can go further in the workflow
+     *
+     * @param string $workflowProcessId
+     * @return bool
+     */
+    public function isNextStepRunnable(string $workflowProcessId) : bool
```
Comment (jrmbrgs) [inline Library/Services/WorkflowManager.php]: infact it should be named `didCurrentStepSucceeded()`

File: Library/Services/WorkflowManager.php [hunk 4]
```diff
@@ -26,47 +27,262 @@ public function __construct(string $workflowConfigFilePath)
...
+    {
+        $this->getDI()->get('logger')->debug(json_encode(func_get_args()));
+        $stepHash = &$this->getContextStepHashRef($workflowProcessId, $stepCode);
+        if (!isset($stepHash['jobList']) || !isset($stepHash['jobList'][$jobId])) {
+            throw new Exceptions\WorkflowException('Cannot find any job');
+        }
+        $stepHash['jobList'][$jobId]['status'] = $resultHash['status'] ?? self::STATUS_FAILED;
+        $stepHash['jobList'][$jobId]['result'] = $resultHash['data'] ?? [];
+
+        if ($stepHash['jobList'][$jobId]['status'] == self::STATUS_FAILED) {
```
Comment (jrmbrgs) [inline Library/Services/WorkflowManager.php:242]: job failure process should be delegated to a dedicated function.
e.g. :
```PHP
if ($stepHash['jobList'][$jobId]['status'] == self::STATUS_FAILED) {
    $this->processStepJobFailure($workflowProcessId, $stepCode, $jobId, $resultHash);
}

private function processStepJobFailure(string $workflowProcessId, string $stepCode, int $jobId, array $resultHash)
{
   // check step conf to see if the step is "blocking"
   // set the WF status accordingly
}

```

File: Library/Services/WorkflowManager.php [hunk 5]
```diff
@@ -26,47 +27,262 @@ public function __construct(string $workflowConfigFilePath)
...
+     */
+    public function registerStepJob($workflowProcessId, $stepCode, $jobId)
+    {
+        $this->getDI()->get('logger')->debug(json_encode(func_get_args()));
+        // q&d search in context the job for which saving the result
+        $stepHash = &$this->getContextStepHashRef($workflowProcessId, $stepCode);
+        if (!isset($stepHash['jobList'])) {
+            $stepHash['jobList'] = [];
+        }
+        $stepHash['jobList'][] = [
```
Comment (jrmbrgs) [inline Library/Services/WorkflowManager.php]: maybe we should check if the givenjob has not been already registred, is this case, throw an Exc

File: Library/Tasks/ManagerTask.php [hunk 6]
```diff
@@ -32,43 +33,63 @@ protected function initWorker(array $paramHash)
...
-            break;
-        default :
-            $this->getDI()->get('logger')->error("ERR : Unknown message type : {$messageDto->getType()}");
+                $status = $this->workflowManagerService->getStatus($messageDto->getId());
+                $this->getDI()->get('logger')->debug("Id {$messageDto->getId()} is '$status'");
+                if ($status == Services\WorkflowManager::STATUS_FAILED) {
+                    throw new WorkflowException("Id failed {$messageDto->getId()}");
+                }
+
+                if ($this->workflowManagerService->isNextStepRunnable($messageDto->getId()))
```
Comment (jrmbrgs) [inline Library/Tasks/ManagerTask.php]: c.f. didCurrentStepSucceeded comment

execflow should looks like this
```
 - getCurrentStepStatus
   - Success : hasNextStep
      - Y : runNextStep
      - N : finalizeWorkflowExec
   - Failed : processStepFailure
   - Running : noop
```
then we must handle the "end of the workflow execution"

File: Library/Tasks/StepTask.php [hunk 7]
```diff
@@ -34,18 +34,21 @@ protected function usage()
...
+                'id' => $messageDto->getId(),
                 'type' => Dtos\Message::TYPE_STEP_ACK,
-                'step' => $this->topicName,
+                'stepCode' => $messageDto->getStepCode(),
+                'jobId' => $messageDto->getJobId(),
                 'result' => json_encode($resultHash)
-            ])
+            ]
         );
         $this->sendMessage('disturb-' . $this->workflowConfig['name']  . '-manager', $msgDto);
```
Comment (mventuravoyageprive) [inline Library/Tasks/StepTask.php]: Please rebase, there is function to build name

**Pull request merged**